zstackctl/zstackctl/ctl.py

```
@@ -8911,14 +8927,19 @@ def stop_node():
                 else:
                     webapp_dir = os.path.dirname(ctl.zstack_home)
 
+                # unzip to temp dir first, only remove old after success
+                tmp_new_zstack = os.path.join(webapp_dir, 'zstack_upgrade_tmp')
+                linux.rm_dir_force(tmp_new_zstack)
                 shell('cp %s %s' % (new_war.path, webapp_dir))
-                ShellCmd('unzip %s -d zstack' % os.path.basename(new_war.path), workdir=webapp_dir)()
+                ShellCmd('unzip %s -d %s' % (os.path.basename(new_war.path), os.path.basename(tmp_new_zstack)),
+                         workdir=webapp_dir)()
+                linux.rm_dir_force(ctl.zstack_home)
+                os.rename(tmp_new_zstack, ctl.zstack_home)
```

Ruff (0.15.7)

[error] `os` may be undefined, or defined from star imports

(F405)